src/ipld/cid_hashset.rs

```diff
@@ -17,23 +17,139 @@ use cid::Cid;
     pub fn insert(&mut self, cid: Cid) -> bool {
-        match cid.try_into() {
-            Ok(CidVariant::V1DagCborBlake2b(bytes)) => self.v1_dagcbor_blake2b.insert(bytes),
-            Err(()) => self.fallback.insert(cid),
+        if self.0.contains_key(cid) {
+            false
+        } else {
+            self.0.insert(cid, ()).is_none()
         }
     }
 
+    /// Returns the number of items in the set.
     pub fn len(&self) -> usize {
-        self.v1_dagcbor_blake2b.len() + self.fallback.len()
+        self.0.len()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct CidHashMap<V> {
+    v1_dagcbor_blake2b_hash_map: HashMap<[u8; BLAKE2B256_SIZE], V>,
+    fallback_hash_map: HashMap<Cid, V>,
+}
+
+impl<V> CidHashMap<V> {
+    /// Creates an empty `HashMap` with CID type keys.
+    pub fn new() -> Self {
+        Self {
+            v1_dagcbor_blake2b_hash_map: HashMap::new(),
+            fallback_hash_map: HashMap::new(),
+        }
+    }
+
+    /// Returns `true` if the map contains a value for the specified key.
+    pub fn contains_key(&self, k: Cid) -> bool {
+        match k.try_into() {
+            Ok(CidVariant::V1DagCborBlake2b(bytes)) => {
+                self.v1_dagcbor_blake2b_hash_map.contains_key(&bytes)
+            }
+            Err(()) => self.fallback_hash_map.contains_key(&k),
+        }
+    }
+
+    /// Inserts a key-value pair into the map; if the map did not have this key present, [`None`] is returned.
+    pub fn insert(&mut self, k: Cid, v: V) -> Option<V> {
+        match k.try_into() {
+            Ok(CidVariant::V1DagCborBlake2b(bytes)) => {
+                self.v1_dagcbor_blake2b_hash_map.insert(bytes, v)
+            }
+            Err(()) => self.fallback_hash_map.insert(k, v),
+        }
+    }
+
+    /// Removes a key from the map, returning the value at the key if the key
+    /// was previously in the map.
+    pub fn remove(&mut self, k: Cid) -> Option<V> {
+        match k.try_into() {
+            Ok(CidVariant::V1DagCborBlake2b(bytes)) => {
+                self.v1_dagcbor_blake2b_hash_map.remove(&bytes)
+            }
+            Err(()) => self.fallback_hash_map.remove(&k),
+        }
+    }
+
+    /// Returns the number of elements the map can hold without reallocating.
+    pub fn capacity(&self) -> usize {
+        self.v1_dagcbor_blake2b_hash_map.capacity() + self.fallback_hash_map.capacity()
+    }
+
+    /// Returns a reference to the value corresponding to the key.
+    pub fn get(&self, k: Cid) -> Option<&V> {
+        match k.try_into() {
+            Ok(CidVariant::V1DagCborBlake2b(bytes)) => self.v1_dagcbor_blake2b_hash_map.get(&bytes),
+            Err(()) => self.fallback_hash_map.get(&k),
+        }
+    }
+
+    /// Returns the number of elements in the map.
+    pub fn len(&self) -> usize {
+        self.v1_dagcbor_blake2b_hash_map.len() + self.fallback_hash_map.len()
     }
 }
 
 #[cfg(test)]
-mod tests {}
+mod tests {
```

Comment:
Sets and maps usually have more than a single element. Is there a way we can test the code more broadly? Can we use arbitrary sets and maps?